File: spec/frontend/vue_shared/components/crud_component_spec.js
import { nextTick } from 'vue';
import { GlButton, GlIcon, GlBadge, GlAnimatedChevronLgDownUpIcon } from '@gitlab/ui';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import CrudComponent from '~/vue_shared/components/crud_component.vue';
import { useLocalStorageSpy } from 'helpers/local_storage_helper';
import { parseBoolean } from '~/lib/utils/common_utils';

describe('CRUD Component', () => {
  useLocalStorageSpy();

  let wrapper;

  const createComponent = (propsData, slots = {}) => {
    wrapper = shallowMountExtended(CrudComponent, {
      propsData: {
        title: 'CRUD Component title',
        ...propsData,
      },
      scopedSlots: {
        ...slots,
      },
      stubs: { GlButton, GlIcon, GlBadge },
    });
  };

  const findTitle = () => wrapper.findByTestId('crud-title');
  const findDescription = () => wrapper.findByTestId('crud-description');
  const findCount = () => wrapper.findByTestId('crud-count');
  const findIcon = () => wrapper.findComponent(GlIcon);
  const findFormToggle = () => wrapper.findByTestId('crud-form-toggle');
  const findActions = () => wrapper.findByTestId('crud-actions');
  const findForm = () => wrapper.findByTestId('crud-form');
  const findSkeletonLoader = () => wrapper.findByTestId('crud-loading');
  const findEmpty = () => wrapper.findByTestId('crud-empty');
  const findBody = () => wrapper.findByTestId('crud-body');
  const findFooter = () => wrapper.findByTestId('crud-footer');
  const findPagination = () => wrapper.findByTestId('crud-pagination');
  const findCollapseToggle = () => wrapper.findByTestId('crud-collapse-toggle');
  const findChevronIcon = () => wrapper.findComponent(GlAnimatedChevronLgDownUpIcon);

  afterEach(() => {
    localStorage.clear();
  });

  it('renders title', () => {
    createComponent();

    expect(findTitle().text()).toBe('CRUD Component title');
  });

  it('renders the element specified by the containerTag prop', () => {
    createComponent({ containerTag: 'span' });
    expect(wrapper.element.tagName).toBe('SPAN');
  });

  it('renders `title` slot', () => {
    createComponent({}, { title: '<p>Title slot</p>' });

    expect(findTitle().text()).toBe('Title slot');
  });

  it('renders description', () => {
    createComponent({ description: 'Description' });

    expect(findDescription().text()).toBe('Description');
  });

  it('renders `description` slot', () => {
    createComponent({}, { description: '<p>Description slot</p>' });

    expect(findDescription().text()).toBe('Description slot');
  });

  it('renders count and icon appropriately when count does not exist', () => {
    createComponent({ icon: 'rocket' });

    expect(findCount().text()).toBe('0');
    expect(findIcon().props('name')).toBe('rocket');
  });

  it('renders count and icon appropriately when count exists', () => {
    createComponent({ count: 99, icon: 'rocket' });

    expect(findCount().text()).toBe('99');
    expect(findIcon().props('name')).toBe('rocket');
  });

  it('renders `count` slot', () => {
    createComponent({}, { count: '<p>Count slot</p>' });

    expect(findCount().text()).toBe('Count slot');
  });

  it('renders `actions` slot', () => {
    createComponent({}, { actions: '<p>Actions slot</p>' });

    expect(findActions().text()).toBe('Actions slot');
  });

  it('renders and shows `form` slot', async () => {
    createComponent({ toggleText: 'Form action toggle' }, { form: '<p>Form slot</p>' });

    expect(findForm().exists()).toBe(false);
    expect(findFormToggle().text()).toBe('Form action toggle');

    findFormToggle().vm.$emit('click');
    await nextTick();

    expect(findFormToggle().exists()).toBe(false);
    expect(findForm().text()).toBe('Form slot');
  });

  it("doesn't render content while loading", () => {
    createComponent({ isLoading: true }, { default: '<p>Body slot</p>' });

    expect(findSkeletonLoader().exists()).toBe(true);
    expect(findCount().exists()).toBe(false);
    expect(findIcon().exists()).toBe(false);
    expect(findBody().text()).toBe('');
  });

  it('renders `empty` slot', () => {
    createComponent({}, { empty: '<span>Empty message</span>' });

    expect(findSkeletonLoader().exists()).toBe(false);
    expect(findEmpty().text()).toBe('Empty message');
  });

  it('renders `body` slot', () => {
    createComponent({}, { default: '<p>Body slot</p>' });

    expect(findSkeletonLoader().exists()).toBe(false);
    expect(findBody().text()).toBe('Body slot');
  });

  it('renders `footer` slot', () => {
    createComponent({}, { footer: '<p>Footer slot</p>' });

    expect(findFooter().text()).toBe('Footer slot');
  });

  it('renders `pagination` slot', () => {
    createComponent({}, { default: '<p>Body slot</p>', pagination: '<p>Pagination slot</p>' });

    expect(findPagination().text()).toBe('Pagination slot');
  });

  describe('with persistCollapsedState=true', () => {
    describe('when the localStorage key is undefined', () => {
      beforeEach(() => {
        createComponent(
          {
            isCollapsible: true,
            persistCollapsedState: true,
            anchorId: 'test-anchor',
            toggleText: 'Form action toggle',
          },
          { default: '<p>Body slot</p>' },
        );
      });

      it('the collapsible area is not collapsed initially', () => {
        expect(findBody().text()).toBe('Body slot');
      });

      it('toggles the collapsible area and sets the localStorage key to true', async () => {
        findCollapseToggle().vm.$emit('click');
        await nextTick();

        expect(localStorage.setItem).toHaveBeenCalledWith('crud-collapse-test-anchor', true);
        expect(findBody().exists()).toBe(false);
      });

      it('does not emit an event on mounted when no local storage key is set', () => {
        expect(wrapper.emitted('collapsed')).toBeUndefined();
        expect(wrapper.emitted('expanded')).toBeUndefined();
      });
    });

    describe('when the localStorage key is true', () => {
      beforeEach(() => {
        localStorage.setItem('crud-collapse-test-anchor', 'true');
        createComponent(
          {
            isCollapsible: true,
            persistCollapsedState: true,
            anchorId: 'test-anchor',
            toggleText: 'Form action toggle',
          },
          { default: '<p>Body slot</p>' },
        );
      });

      it('the collapsible area is collapsed initially', () => {
        expect(findBody().exists()).toBe(false);
      });

      it('emits an event on mounted', () => {
        expect(wrapper.emitted('collapsed').at(0)).toEqual([]);
        expect(wrapper.emitted('expanded')).toBeUndefined();
      });

      it('toggles the collapsible area and sets the localStorage key to false', async () => {
        findCollapseToggle().vm.$emit('click');
        await nextTick();

        expect(localStorage.setItem).toHaveBeenCalledWith('crud-collapse-test-anchor', false);
        expect(findBody().text()).toBe('Body slot');
      });
    });

    describe('when the localStorage key is false', () => {
      beforeEach(() => {
        localStorage.setItem('crud-collapse-test-anchor', 'false');
        createComponent(
          {
            isCollapsible: true,
            persistCollapsedState: true,
            anchorId: 'test-anchor',
            toggleText: 'Form action toggle',
          },
          { default: '<p>Body slot</p>' },
        );
      });

      it('emits an event on mounted', () => {
        expect(wrapper.emitted('collapsed')).toBeUndefined();
        expect(wrapper.emitted('expanded').at(0)).toEqual([]);
      });
    });
  });

  describe('isCollapsible', () => {
    it('renders collapsible toggle', () => {
      createComponent({ isCollapsible: true }, { default: '<p>Body slot</p>' });

      expect(findCollapseToggle().exists()).toBe(true);
      // Vue compat doesn't know about component props if it extends other component
      expect(
        findChevronIcon().props('isOn') ?? parseBoolean(findChevronIcon().attributes('is-on')),
      ).toBe(true);
    });

    it('click on toggle hides content', async () => {
      createComponent({ isCollapsible: true }, { default: '<p>Body slot</p>' });

      expect(findBody().exists()).toBe(true);

      await findCollapseToggle().vm.$emit('click');

      expect(findBody().exists()).toBe(false);
      // Vue compat doesn't know about component props if it extends other component
      expect(
        findChevronIcon().props('isOn') ?? parseBoolean(findChevronIcon().attributes('is-on')),
      ).toBe(false);
    });

    it('`collapsed` hides content by default', () => {
      createComponent({ isCollapsible: true, collapsed: true }, { default: '<p>Body slot</p>' });

      expect(findBody().exists()).toBe(false);
    });

    it('emits `expanded` when clicked on a collapsed toggle', async () => {
      createComponent({ isCollapsible: true, collapsed: true }, { default: '<p>Body slot</p>' });

      await findCollapseToggle().vm.$emit('click');

      expect(wrapper.emitted('expanded')).toStrictEqual([[]]);
    });

    it('emits `click-expanded` when clicked on a collapsed toggle, allowing for click tracking', async () => {
      createComponent({ isCollapsible: true, collapsed: true }, { default: '<p>Body slot</p>' });

      await findCollapseToggle().vm.$emit('click');

      expect(wrapper.emitted('click-expanded')).toStrictEqual([[]]);
    });

    it('emits `collapsed` when clicked on an expanded toggle', async () => {
      createComponent({ isCollapsible: true }, { default: '<p>Body slot</p>' });

      await findCollapseToggle().vm.$emit('click');

      expect(wrapper.emitted('collapsed')).toStrictEqual([[]]);
    });

    it('emits `click-collapsed` when clicked on an expanded toggle, allowing for click tracking', async () => {
      createComponent({ isCollapsible: true }, { default: '<p>Body slot</p>' });

      await findCollapseToggle().vm.$emit('click');

      expect(wrapper.emitted('click-collapsed')).toStrictEqual([[]]);
    });
  });

  describe('default slot', () => {
    it('passes the showForm function to the default slot', () => {
      const defaultSlot = jest.fn();
      createComponent({}, { default: defaultSlot });

      expect(defaultSlot).toHaveBeenCalledWith(
        expect.objectContaining({ showForm: wrapper.vm.showForm }),
      );
    });
  });

  describe('actions slot', () => {
    it('passes the showForm function to the actions slot', () => {
      const actionsSlot = jest.fn();
      createComponent({}, { actions: actionsSlot });

      expect(actionsSlot).toHaveBeenCalledWith(
        expect.objectContaining({ showForm: wrapper.vm.showForm }),
      );
    });
  });

  describe('showZeroCount prop', () => {
    it('displays "0" when count is falsy and showZeroCount is true', () => {
      createComponent({ showZeroCount: true, count: null });

      expect(findCount().text()).toBe('0');
    });

    it('does not display count when count is falsy and showZeroCount is false', () => {
      createComponent({ showZeroCount: false, count: null });

      expect(findCount().exists()).toBe(false);
    });

    it('displays "0" as string when count is 0 and showZeroCount is true', () => {
      createComponent({ showZeroCount: true, count: 0 });

      expect(findCount().text()).toBe('0');
    });

    it('displays regular count when count exists regardless of showZeroCount', () => {
      createComponent({ showZeroCount: true, count: 5 });

      expect(findCount().text()).toBe('5');
    });
  });
});

File: spec/components/layouts/crud_component_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe Layouts::CrudComponent, type: :component, feature_category: :shared do
  let(:title) { 'Title' }
  let(:description) { 'Description' }
  let(:count) { 99 }
  let(:icon) { 'rocket' }
  let(:toggle_text) { 'Toggle text' }
  let(:actions) { 'Actions' }
  let(:body) { 'Body' }
  let(:form) { 'Form' }
  let(:footer) { 'Footer' }
  let(:pagination) { 'Pagination' }
  let(:component_title) { described_class.new(title) }

  describe 'slots' do
    it 'renders title' do
      render_inline component_title

      expect(page).to have_css('[data-testid="crud-title"]', text: title)
    end

    it 'renders description' do
      render_inline described_class.new(title, description: description)

      expect(page).to have_css('[data-testid="crud-description"]', text: description)
    end

    it 'renders description slot' do
      render_inline component_title do |c|
        c.with_description { description }
      end

      expect(page).to have_css('[data-testid="crud-description"]', text: description)
    end

    it 'renders count and icon' do
      render_inline described_class.new(title, count: count, icon: icon)

      expect(page).to have_css('[data-testid="crud-count"]', text: count)
      expect(page).to have_css('[data-testid="crud-count"] svg[data-testid="rocket-icon"]')
    end

    it 'renders icon class' do
      render_inline described_class.new(title, count: count, icon: icon, icon_class: 'test-icon-class')

      expect(page).to have_css('[data-testid="crud-count"] svg[data-testid="rocket-icon"].test-icon-class')
    end

    it 'renders action toggle' do
      render_inline described_class.new(title, toggle_text: toggle_text)

      expect(page).to have_css('[data-testid="crud-action-toggle"]', text: toggle_text)
      expect(page).to have_css('.crud.js-toggle-container')
      expect(page).to have_css('[data-testid="crud-action-toggle"].js-toggle-button.js-toggle-content')
    end

    it 'renders action toggle custom attributes' do
      render_inline described_class.new(title,
        toggle_text: toggle_text,
        toggle_options: { class: 'custom-button-class', data: { testid: 'crud-custom-toggle-id' } })

      expect(page).to have_css('.custom-button-class', text: toggle_text)
      expect(page).to have_css('[data-testid="crud-custom-toggle-id"]', text: toggle_text)
    end

    it 'renders actions slot' do
      render_inline component_title do |c|
        c.with_actions { actions }
      end

      expect(page).to have_css('[data-testid="crud-actions"]', text: actions)
    end

    it 'renders form slot' do
      render_inline component_title do |c|
        c.with_form { form }
      end

      expect(page).to have_css('[data-testid="crud-form"]', text: form)
    end

    it 'renders hidden form slot if toggle is set' do
      render_inline described_class.new(title, toggle_text: toggle_text) do |c|
        c.with_form { form }
      end

      expect(page).to have_css('.gl-hidden', text: form)
      expect(page).to have_css('[data-testid="crud-form"].js-toggle-content', text: form)
    end

    it 'renders form visible when form has errors and toggle_text is present' do
      render_inline described_class.new(title,
        toggle_text: toggle_text,
        form_options: { form_errors: true }) do |c|
        c.with_form { form }
      end

      expect(page).not_to have_css('.gl-hidden', text: form)
    end

    it 'renders form custom attributes' do
      render_inline described_class.new(title,
        form_options: { class: 'form-class', data: { testid: 'crud-custom-form-id' } }) do |c|
        c.with_form { form }
      end

      expect(page).to have_css('.form-class', text: form)
      expect(page).to have_css('[data-testid="crud-custom-form-id"]', text: form)
    end

    it 'renders body slot' do
      render_inline component_title do |c|
        c.with_body { body }
      end

      expect(page).to have_css('[data-testid="crud-body"]', text: body)
    end

    it 'renders body custom attributes' do
      render_inline described_class.new(title,
        body_options: { class: '!gl-m-0', data: { testid: 'crud-custom-body-id' } }) do |c|
        c.with_body { body }
      end

      expect(page).to have_css('.\!gl-m-0', text: body)
      expect(page).to have_css('[data-testid="crud-custom-body-id"]', text: body)
    end

    it 'renders footer slot' do
      render_inline component_title do |c|
        c.with_footer { footer }
      end

      expect(page).to have_css('[data-testid="crud-footer"]', text: footer)
    end

    it 'renders footer custom attributes' do
      render_inline described_class.new(title,
        footer_options: { class: 'footer-class', data: { testid: 'crud-custom-footer-id' } }) do |c|
        c.with_footer { footer }
      end

      expect(page).to have_css('.footer-class', text: footer)
      expect(page).to have_css('[data-testid="crud-custom-footer-id"]', text: footer)
    end

    it 'renders pagination slot' do
      render_inline component_title do |c|
        c.with_pagination { pagination }
      end

      expect(page).to have_css('[data-testid="crud-pagination"]', text: pagination)
    end
  end

  describe 'collapsible functionality' do
    it 'renders collapse/expand button when is_collapsible is true' do
      render_inline described_class.new(title, is_collapsible: true)

      expect(page).to have_css('.js-crud-collapsible-button', text: '')
      expect(page).to have_css('.js-crud-collapsible-button[aria-expanded="true"]')
      expect(page).to have_css('.js-crud-collapsible-button[aria-controls]')
      expect(page).to have_css('.js-crud-collapsible-button[title="Collapse"]')
      expect(page).to have_css('.js-crud-collapsible-button[data-collapse-title="Expand"]')
      expect(page).to have_css('.js-crud-collapsible-button[data-expand-title="Collapse"]')
      expect(page).to have_css('.js-crud-collapsible-collapse')
      expect(page).to have_css('.js-crud-collapsible-expand.gl-hidden')
    end
  end
end
